src/core/qgsvectorlayer.cpp

```diff
@@ -692,15 +692,22 @@ long QgsVectorLayer::featureCount( const QString &legendKey ) const
   return mSymbolFeatureCountMap.value( legendKey );
 }
 
+const QgsFeatureIds QgsVectorLayer::featureIds( const QString &legendKey ) const
+{
+  if ( !mSymbolFeatureCounted )
+    return QgsFeatureIds();
 
+  return mSymbolIdMap.value( legendKey, QgsFeatureIds() );
+}
 
-QgsVectorLayerFeatureCounter *QgsVectorLayer::countSymbolFeatures()
+QgsVectorLayerFeatureCounter *QgsVectorLayer::countSymbolFeatures( bool disable_async )
```

Comment:
disable_async looks like something we should try to avoid.

* it will block the main thread, potentially for a long time
* this shouldn't be run on any other thread than the main thread or we risk crashes

Reply:
I agree, but otherwise with the tests purely in python counting is never done, is suspect this is due to a lack of app to handle the background task.

Reply:
What you can do is something like

```
loop = QEventLoop()
counterTask.featuresCounted.connect(loop.quit)
# start counting features here
loop.exec_()
```

Reply:
Maybe? I'm not to sure where this would go and how to trigger it only in python mode without the app. Some pointers would help.

Reply:
This should go into the tests, whenever you are waiting for a response from a background task in a test/standalone app, this is how to wait for it.

Reply:
But where does the counterTask comes from? Do I need to extract it from the countSymbolFeature or is this something else?

Reply:
Exactly, arount these lines here https://github.com/qgis/QGIS/pull/9648/files?file-filters%5B%5D=.py#diff-97664ef6f393f333c7d7c2ef4f68e047R383

Reply:
doesn't extracting the task entails even more drastic api changes?